🐛 `TrajectoryData`: Add `pbc`

mbercx opened the pull request:
>[!NOTE]
> There is a lot of code debt in the `TrajectoryData`, and the API is quite old and not optimal. Fixing all of that is beyond the scope of this PR, and probably we should redesign it in `aiida-atomistic`.

Fixes https://github.com/aiidateam/aiida-core/issues/6376
Fixes https://github.com/aiidateam/aiida-core/issues/7037

Pinging @danielhollas, @rikigigi and @dbidoggia for review, since they raised the original issues.

The current changes try to remain backwards-compatible for `get_step_data`, i.e. it now keeps on returning the same tuple of length 5. However, this does mean that the method does not return the `pbc`. For the `get_step_structure` method, I directly set the `pbc` from the property obtained from the attributes.

danielhollas commented on src/aiida/orm/nodes/data/array/trajectory.py:
```diff
@@ -131,14 +137,20 @@ def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=Non
 
-        self._internal_validate(stepids, cells, symbols, positions, times, velocities)
-        # set symbols as attribute for easier querying
+        pbc = pbc or [True, True, True]
```
```suggestion
        pbc = pbc or [True, True, True]
```
This default makes me uneasy. Should we instead look at the `cells` and see if it is defined and non-zero?

mbercx replied:
Sensible check! My MC3D bias is showing again ^^

mbercx replied:
> if it is defined and non-zero?

Can the `cells` be defined as zero? 🤔

dbidoggia replied:
Do not know when I would use a structure without specifying cells, but agree to consider them `False` if not specified

mbercx replied:
Apparently you can set e.g. `cell = [[0, 0, 0], [0, 1, 0], [0, 0, 1]]`, but I'm not sure that's sensible? @danielhollas right now I've adapted the default to be `[False, False, False]` in case `cells=None`. Is that sufficient? Or what did you have in mind?

danielhollas reviewed (changes requested):
@mbercx thank you very much for working on this. I'd like to also try this out in my app to verify.

Also, can you please add test cases for the default PBC behaviour?

mbercx commented:
@danielhollas if you have a moment, could you have another look? We could still squeeze this into 2.7.2, since it's fixing a bug in my view. I want to avoid having this open too long and it falling through the cracks. 🙏

danielhollas reviewed (dismissed):
@mbercx just a quick note that I haven't forgotten about this, and plan to test this as part of the coding week. :-) 

Had a quick look and the code looks great, just some minor suggestions. 

Also, do we need to update documentation somewhere? (I haven't checked)

I haven't gone through the tests yet.

mbercx commented:
Thanks @danielhollas! Round 3 done, let me know if there is anything else.

> Also, do we need to update documentation somewhere? (I haven't checked)

`TrajectoryData` is rather [succinctly documented](https://aiida.readthedocs.io/projects/aiida-core/en/stable/topics/data_types.html#trajectorydata) I'm afraid, but I do not intend to fix that here.

danielhollas reviewed (changes requested):
Really sorry for stalling this for so long. Finally got a chance to test this in AiiDAlab app and I did run into an issue. When I load a pre-existing TrajectoryData from DB, calling `get_step_structure` will raise an attribute error.

```python

~/.local/lib/python3.9/site-packages/aiidalab_widgets_base/utils/__init__.py in get_formula(data_node)
    116         # so this approach is sound.
    117         stepid = data_node.get_stepids()[0]
--> 118         return data_node.get_step_structure(stepid).get_formula()
    119     elif isinstance(data_node, StructureData):
    120         return data_node.get_formula()

~/aiida-core/src/aiida/orm/nodes/data/array/trajectory.py in get_step_structure(self, index, custom_kinds)
    423                 )
    424 
--> 425         struc = StructureData(cell=cell, pbc=self.pbc)
    426         if custom_kinds is not None:
    427             for _k in custom_kinds:

~/aiida-core/src/aiida/orm/nodes/node.py in __getattr__(self, name)
    839             return getattr(self.base.links, new_name)
    840 
--> 841         raise AttributeError(name)

AttributeError: pbc
```

I am unsure how this should be handled. How is adding a new attribute handled generally?? Is there some DB migration needed? Or just a code adjustment to account for this attribute possibly missing?

GeigerJ2 commented:
Hi guys, just wondering what's the status here? Is this feasible soon, or should I move it out of v2.8 and into v2.9? Thanks!

danielhollas commented on src/aiida/orm/nodes/data/array/trajectory.py:
```diff
@@ -131,14 +140,28 @@ def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=Non
-        self._internal_validate(stepids, cells, symbols, positions, times, velocities)
-        # set symbols as attribute for easier querying
+        if cells is None:
+            pbc = pbc or [False, False, False]
```
We're converting to tuple below so might as well use tuple here


```suggestion
            pbc = pbc or (False, False, False)
```

danielhollas commented on src/aiida/orm/nodes/data/array/trajectory.py:
```diff
@@ -82,8 +85,14 @@ def _internal_validate(self, stepids, cells, symbols, positions, times, velociti
+            raise ValueError('Periodic boundary conditions are only possible when a cell is defined.')
+
+    def set_trajectory(
+        self, symbols, positions, stepids=None, cells=None, times=None, velocities=None, pbc: None | list | tuple = None
```
```suggestion
        self, symbols, positions, stepids=None, cells=None, times=None, velocities=None, pbc: None | list[bool] | tuple[bool, bool, bool] = None
```

danielhollas reviewed (approved):
Okay, I've retested the current version and verified (in my AiiDAlab app) that the code works for TrajectoryData without pbc attribute. Thanks for the fix @mbercx 👍 

I am approving since this all looks great, I just have some final minor suggestions and typing fixes.

@mbercx since this file is currently not type-checked via pre-commit, can you run mypy manually?

```console
$ mypy --pretty src/aiida/orm/nodes/data/array/trajectory.py
```

You'll likely get a bunch of errors, but the number of errors should not be greater than on main. :-) 

Thanks for sticking with this and sorry for being slow to review.

Oh, it would be great to document this (not sure if TrajectoryData has a dedicated documentation?) and perhaps write down a CHANGELOG entry to help @GeigerJ2 with a release. :-)

mbercx commented on tests/orm/nodes/data/test_trajectory.py:
```diff
@@ -123,6 +127,8 @@ def test_trajectory_get_step_data_empty(self, trajectory_data):
         assert np.array_equal(positions, trajectory_data['positions'][3, :, :])
         assert velocities is None
+        # In case the cell is not defined, there should be no periodic boundary conditions
+        assert np.array_equal(trajectory.pbc, [False, False, False])
```
Hehe, sure. I just copied the line and adapted.

mbercx committed:
🐛 `TrajectoryData`: Fix `pbc` handling for non-periodic structures

Currently, the `TrajectoryData` is not functional for structures that are not periodic,
since their `pbc` property is not considered when using the `set_structurelist` method,
which is also used in the constructor.

Here we add the `pbc` attribute to the `TrajectoryData` class, which is set in the
`set_trajectory` method. The value should be a list/tuple of length 3 with boolean
values, which is validated in the `_internal_validate` method.  The `pbc` should also be
the same for all structures in the trajectory, which is validated in the
`set_structurelist` method. In line with the current approach for `symbols`, a property
is added. This property is used in the `get_step_structure` method to set the `pbc` in
the `StructureData` constructor. The pydantic model is also updated with a new `pbc`
field.

The current changes try to remain backwards-compatible:

1. Since old `TrajectoryData` will not have the `pbc` AiiDA attribute in the database,
   the `pbc` property returns `None` in case it is missing. We also considered doing a
   migration to add the `pbc` attribute, but it is unclear what is the correct default,
   and migrations are often disruptive for the user.
2. The `pbc` input is optional in the `set_trajectory` method. In case it is not
   provided, the structure is assumed to be periodic for trajectories which define
   `cells`, and aperiodic in all directions otherwise. However, not specifying the
   `pbc` when `cells` are defined is deprecated and will raise in v3.0.
3. The `get_step_data` method still returns the same tuple of length 5. However, this
   does mean that the method does not return the `pbc`.
